libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs

```diff
@@ -1,25 +1,177 @@
+bootstrapDStateSpec ::
+  IsConwayUniv fn =>
+  CertsContext Conway ->
+  Specification fn (DState Conway)
+bootstrapDStateSpec withdrawals =
+  let isKey (ScriptHashObj _) = False
+      isKey (KeyHashObj _) = True
+      withdrawalPairs = Map.toList (Map.mapKeys snd (Map.map coinToWord64 withdrawals))
+      withdrawalKeys = Map.keysSet (Map.mapKeys snd withdrawals)
+   in constrained $ \ [var| dstate |] ->
+        match dstate $ \ [var| rewardMap |] _futureGenDelegs _genDelegs _rewards ->
+          [ assert $ sizeOf_ _futureGenDelegs ==. 0
+          , match _genDelegs $ \gd -> assert $ sizeOf_ gd ==. 0
+          , match _rewards $ \w x y z -> [sizeOf_ w ==. 0, sizeOf_ x ==. 0, y ==. lit mempty, z ==. lit mempty]
+          , match [var| rewardMap |] $ \ [var| rdMap |] [var| ptrMap |] [var| sPoolMap |] dRepDelegs ->
+              [ assertExplain (pure "dom sPoolMap is a subset of dom rdMap") $ dom_ sPoolMap `subset_` dom_ rdMap
+              , assertExplain (pure "dom ptrMap is empty") $ dom_ ptrMap ==. mempty
+              , assertExplain (pure "some rewards (not in withdrawals) are zero") $
+                  forAll rdMap $
+                    \ [var| keycoinpair |] -> match keycoinpair $ \cred [var| rdpair |] ->
+                      -- Apply this only to entries NOT IN the withdrawal set, since withdrawals already set the reward in the RDPair.
+                      whenTrue (not_ (member_ cred (lit withdrawalKeys))) (satisfies rdpair someZeros)
+              , forAll (lit (Set.filter isKey withdrawalKeys)) $ \cred -> assert $ member_ cred (dom_ dRepDelegs)
+              , forAll (lit withdrawalPairs) $ \ [var| pair |] ->
+                  match pair $ \ [var| cred |] [var| coin |] ->
+                    [ assert $ member_ cred (dom_ rdMap)
+                    , (caseOn (lookup_ cred rdMap))
+                        -- Nothing
+                        ( branch $ \_ -> FalsePred (pure ("credential " ++ show cred ++ " not in rdMap, bootstrapCertStateSpec"))
+                        )
+                        -- Just
+                        ( branch $ \ [var| rdpair |] ->
+                            match rdpair $ \rew _deposit -> assert $ rew ==. coin
+                        )
+                    ]
+              ]
+          ]
+
+coinToWord64 :: Coin -> Word64
+coinToWord64 (Coin n) = fromIntegral n
+
+type CertsContext era = (Map (Network, Credential 'Staking (EraCrypto era)) Coin)
+
+txZero :: AlonzoTx Conway
+txZero = AlonzoTx mkBasicTxBody mempty (IsValid True) def
 
 certsEnvSpec ::
   IsConwayUniv fn =>
-  Specification fn (CertsEnv (ConwayEra StandardCrypto))
-certsEnvSpec = constrained $ \env ->
-  match env $ \_ pp _ _ _ _ ->
-    satisfies pp pparamsSpec
+  Specification fn (CertsEnv Conway)
+certsEnvSpec = constrained $ \ce ->
+  match ce $ \tx pp _ _ a b ->
+    [ satisfies pp pparamsSpec
+    , assert $ tx ==. lit txZero
+    , assert $ a ==. lit SNothing
+    , assert $ sizeOf_ b ==. 0
```

Comment:
Overconstrained

Reply:
Loosened it a bit.

match ce $ \tx pp _slot _currepoch _currcommittee commproposals ->
    [ satisfies pp pparamsSpec
    , assert $ tx ==. lit txZero
    , genHint 3 commproposals
    ]